Add SAML IdP plugin for internal use

tvernum opened the pull request:
This change merges the "feature-internal-idp" branch into Elasticsearch.

This introduces a small `identity-provider` plugin as a child of the `x-pack` module.
This allows ES to act as a SAML IdP, for users who are authenticated against the
Elasticsearch cluster.

This feature is intended for internal use within Elastic Cloud environments
and is not supported for any other use case. It falls under an enterprise license tier.

The IdP is disabled by default.

tvernum committed:
X-Pack SAML IdP Plugin (#51514)

This introduces a new sub-plugin to X-Pack that implements the backend
for a basic IdP built on top of the X-Pack security primitives.

In this commit there is a mostly empty plugin implementation and a
class the implements the minimum necessary to generate a SAML 2.0
authentication "Response".

jkakavas committed:
Merge remote-tracking branch 'origin/master' into feature-internal-idp

jkakavas committed:
Add IDP Configuration settings and related tests (#51682)

jkakavas committed:
Merge remote-tracking branch 'origin/master' into feature-internal-idp

tvernum committed:
Merge branch 'master' into feature-internal-idp

tvernum committed:
Add signing support to SAML IdP (#51517)

Adds a new "SamlObjectSigner" class that can sign messages such as
authn Response objects

tvernum committed:
Merge branch 'master' into feature-internal-idp

tvernum committed:
Merge branch 'master' into feature-internal-idp

jkakavas committed:
IDP-initiated sso REST handler (#51830)

Necessary rest handler and transport action for consuming an SP
EntityID and producing a SAML response for the authenticating user.
TransportSamlInitiateSingleSignOnAction needs to be adjusted to
take advantage of the secondary authentication handling logic once
this is available.

tvernum committed:
Resolve application privileges to SP access (#52250)

Adds a new UserPrivilegeResolver class that determines whether a user
has access to a service, and with what groups, based on an application
resource model that represents the service.

tvernum committed:
Merge branch 'master' into feature-internal-idp

tvernum committed:
Fix compile / merge issue (#52643)

Compilation of the feature-internal-idp branch was broken due to
conflicting merges.

tvernum committed:
Write SAML SP data to a dedicated index (#52401)

This adds low level classes for reading & writing documents that
model a SAML Service Provider in the IdP.
(These classes are not currently used)

tvernum committed:
Automatically install SAML SP index template (#52706)

Triggers the installation of the index template automatically due to
cluster state changes, or before writing a SP doc

tvernum committed:
Update IdP plugin security policy (#52740)

ApacheXMLSecurityInitializer (a dependency of OpenSAML) attempts to
set the "org.apache.xml.security.ignoreLineBreaks" system property if
it doesn't exist

jkakavas committed:
Handle incoming AuthnRequests (#52018)

- Expose an API that consumes (possibly signed) AuthnRequests as
defined by the HTTP-Redirect binding.
- Process AuthnRequests, validate and parse them into a minimum
set of information to be used for subsequent API calls to get a
SAML Response

jkakavas committed:
Create SAML Responses for secondary authN (#52805)

Create the SAML responses for the principal that we should. The
credentials of this user need to have been passed in the
es-secondary-authorization header.

tvernum committed:
Merge branch 'master' into feature-internal-idp

jkakavas committed:
Generate SAML metadata for the Identity Provider (#52415)

This change adds an API for generating SAML metadata for the
configured Identity Provider. All parameters that can contribute
to the metadata, from signing keys to Organization and contact
details are all configured in node settings.
The caller of the API needs to define the SP that will consume this
metadata as the values included can potentially be different,
depending on the registered SP configuration and the IDP policy.

tvernum committed:
Merge branch 'master' into feature-internal-idp

tvernum committed:
Add REST endpoint to register a new SAML SP (#52936)

This change adds a new PUT /_idp/saml/sp/{entity_id} endpoint that
registers (or updates) a Service Provider in the IDP.

tvernum committed:
Merge branch 'master' into feature-internal-idp

tvernum committed:
Make ServiceProvider lookup asynchronous (#52986)

This prepares for reading the real Service Provider data from the saml
index.

We intend to support a large number of ServiceProviders within a
single cluster, so it is not reasible to keep them all cached in
memory. Since the read from the underyling index will be asynchronous,
all methods that make use of a service provider also need to be
asychronous.

tvernum committed:
Resolve runtime SP model from index doc (#52985)

This change adds a `SamlServiceProviderResolver` that can build a
`SamlServiceProvider` object for a given entity-id, based on the
document model stored in `SamlServiceProviderDocument`.

This resolver includes a cache that is aware of document versioning
(seqNo, primaryTerm) and avoids JSON parsing and object construction
if the document has not changed since last use.

tvernum committed:
Refactor SamlIdentityProvider setup (#52990)

The SamlIdentityProvider / CloudIdp classes had evolved gradually from
their initial commit to the point where they no longer were quite the
right fit for the project.

This change merges the class and interface, and migrates some of the
instantiation logic that had been split between the Plugin and the
class into a separate Builder object.

We also move the Contact and Organisation information from the
metadata builder into the IdentityProvider class so that the class is
self-contained.

tvernum committed:
Use optionalVLong (#53149)

Updates SamlServiceProviderDocument to use the new
read/writeOptionalVLong methods from #53145

tvernum committed:
Rename "groups" to "roles" (#53208)

The Identity Provider has a feature to release a standard set of
attributes to a Service Provider. One of those attributes is intended
to represent the groups or roles to which the user belongs.

From an Identity Provider point of view, there is no concrete
definition of what these attributes represent or should be used for -
the Service Provider may treat them as "groups" (a named collection of
users) or "roles" (a named representation of the user's access
rights).

Terminology across IdPs varies, but we have decided that within the ES
IdP, we will refer to them as "roles". This change replaces any use of
the term "groups" with "roles".

jkakavas committed:
Merge remote-tracking branch 'origin/master' into feature-internal-idp

tvernum committed:
Resolve real (indexed) SPs from IdentityProvider (#53211)

This change hooks the SamlIdentityProvider class up to the
SamlServiceProviderResolver so that we return real registered
service providers during SSO handling

tvernum committed:
Merge branch 'master' into feature-internal-idp

tvernum committed:
Resolve correct service privileges for user (#53213)

When a user SSOs to a service provider, the IdP will provide
attributes to indicate the user's access (roles) in that service.

This change connect the component that determines the user's access
(UserPrivilegeResolver) to the action that generates the SAML response
(TransportSamlInitiateSingleSignOnAction).

jkakavas committed:
Merge remote-tracking branch 'origin/master' into feature-internal-idp

jkakavas committed:
Merge remote-tracking branch 'origin/master' into feature-internal-idp

tvernum committed:
Remove explicit login action and application (#53320)

This removes the concept of a login action, and instead relies
exclusively on the existence of roles to determine whether a user has
access to a service resource.

It also removes the ability for a service provider to be configured to have a
resource under a custom application name.

jkakavas committed:
Add missing parameters to HasPrivileges requests (#53396)

The request must contain cluster and indices privileges

jkakavas committed:
Sign SAML responses by default (#53382)

Sign all SAML Responses ( the Response Element without signing
the included Assertion ) that the Idp sends to any SP. To that
effect, make it an error to not set a signing credential when
building a SamlIdentityProvider.

jkakavas committed:
Return appropriate error codes (#53371)

When we fail to validate an Authentication request because the
request is invalid, respond with 400 instead of 500.

tvernum committed:
Merge branch 'master' into feature-internal-idp

tvernum committed:
Configure SP defaults from YAML settings (#53279)

This change adds support for reading the SP defaults from the
elasticsearch Settings

jkakavas committed:
Merge remote-tracking branch 'origin/master' into feature-internal-idp

tvernum committed:
Merge branch 'master' into feature-internal-idp

tvernum committed:
Add endpoint to delete a service provider by ID (#53322)

This API is a sibling to the existing Put SAML Service Provider API
that registers a new SP.

The implementation is a hard delete.

jkakavas committed:
Handle refresh policy for the request to write the SP to the index (#53364)

Set default refresh policy to NONE

jkakavas committed:
Handle SamlAuthenticationState between APIs (#53254)

After successful validation of an AuthnRequest the caller of the
/validate API needs to call /init to generate the SAML response
that it can then send to the SP. There is some context that needs
to be carried from the response of the /validate API to the
request to the /init API. This context is encapsulated in the
SamlAuthenticationState object that is serialized to a json
object in the SamlValidateAuthnRequestResponse. This can be
treated as an opaque blob for the caller of the API that just
needs to pass it to the subsequent request to the /init API. At
the same time we enable SAML responses that indicate failure
(as opposed to returning a 4xx) for both the /init API as
dictated by the SAML 2.0 Web Browser SSO profile.

Additionally these small adjustments were made:

- Saml Service Providers must indicate a single NameID format that
  they want to use upon registration, instead of multiple. This is
  used to validate the NameID Policy of the AuthnRequest if it
  exists or to determine the NameID format for the response
  otherwise.
- If no NameID Policy is set in the AuthnRequest and the SP has
  not registered a desired one, an IDP wide default NameID format
  is returned.
- JSON response fomat for the RestSamlInitiateSingleSignOnAction
  has been reviewed as discussed.
- All Rest actions have been moved under the same package

jkakavas committed:
Merge remote-tracking branch 'origin/master' into feature-internal-idp

tvernum committed:
Merge branch 'master' into feature-internal-idp

jkakavas committed:
Add supported NameID Format configuration and generate transient values (#53593)

Only transient format is supported for now.

jkakavas committed:
Merge remote-tracking branch 'origin/master' into feature-internal-idp

tvernum committed:
Add a QA project for identity provider module (#53585)

This adds a small QA project that starts up a cluster with the IdP
enabled and performs a basic test of registering a service provider.

It also fixes a few bugs that were found along the way.

jkakavas committed:
Populate SAML Response Attribute Values (#53665)

Add name, email and principal as attributes in the successful
SAML response

jkakavas committed:
Merge branch 'feature-internal-idp' of github.com:elastic/elasticsearch into feature-internal-idp

jkakavas committed:
License Checks for IDP (#53651)

Idp related functionality should only be available for enterprise
licenses.

jkakavas committed:
Merge remote-tracking branch 'origin/master' into feature-internal-idp

jkakavas committed:
Merge branch 'master' into feature-internal-idp

jkakavas committed:
Merge remote-tracking branch 'origin/master' into feature-internal-idp

jkakavas committed:
Add Integration tests for the new IDP handlers (#53130)

Adds a base class for IDP related EsIntegTests (IdpIntegTestCase)
and tests for the existing APIs

jkakavas committed:
Fix IdP signing (#53971)

Fixes signing to run with the correct context classloader.
This in turn requires doPrivileged.

tvernum committed:
Merge branch 'master' into feature-internal-idp

elasticmachine commented:
Pinging @elastic/es-security (:Security/Security)

jasontedor commented on test/framework/src/main/java/org/elasticsearch/test/SerializationTestUtils.java:
```diff
@@ -0,0 +1,60 @@
+/**
+ * Utility methods for assisting with testing {@link Writeable} object.
+ */
+public class SerializationTestUtils {
```
This is a drive-by comment. I can't find any uses of this? And also, I think that `AbstractWireSerializingTestCase` provides roughly the same functionality?

tvernum replied:
It's used in 
- idp.action.DeleteSamlServiceProviderRequestTests
- idp.action.PutSamlServiceProviderRequestTests
- idp.saml.sp.SamlServiceProviderDocumentTests
- idp.saml.support.SamlAuthenticationStateTests

You're right, I can switch those to use the existing methods.
I was sure we had something like that, but I just couldn't find it, so I figured I must have remembered incorrectly.

jasontedor replied:
Strange, the IDE showed me there were not any uses. Either way, thanks for switching over.

tvernum committed:
Fix test to always have roles (#54048)

This change fixes a radomised failure in
TransportSamlInitiateSingleSignOnRequestTests

Under some seeds (e.g. 856F19C80D0A73B6) the roles array would be
empty, which would mean that the generated SAML response would not
contain a roles attribute.

tvernum committed:
Remove SerializationTestUtils (#54052)

This change replaces the SerializationTestUtils class with methods
from ESTestCase

jkakavas reviewed (approved):
:shipit: 

(after we merge https://github.com/elastic/elasticsearch/pull/54053 )

jkakavas committed:
Wrap RestorableContextClassLoader with doPrivileged (#54053)

Needed since RestorableContextClassLoader was moved to core (from
security)
